rust/nix-util/src/context.rs

```diff
@@ -34,6 +46,46 @@ impl Context {
         }
         Ok(())
     }
+
+    fn clear(&self) {
+        unsafe {
+            raw::set_err_msg(
+                self.inner.as_ptr(),
+                raw::NIX_OK.try_into().unwrap(),
+                b"\0".as_ptr() as *const i8,
+            );
+        }
+    }
+
+    pub fn check_err_and_clear(&self) -> Result<()> {
+        let r = self.check_err();
+        if r.is_err() {
+            self.clear();
+        }
+        r
+    }
+
+    /// Run the function, and check the error, then reset the error.
+    /// Make at most one call to a Nix function in `f`.
+    /// Do not use if the context isn't fresh or cleared (e.g. with `check_err_and_clear`).
+    pub fn check_one_call<T, F: FnOnce(*mut raw::c_context) -> T>(&self, f: F) -> Result<T> {
```

Comment:
I didn't go all the way down the road of polishing this, but this macro does basically what I was imagining
```rust
#[macro_export]
macro_rules! check_call {
    ($f:path[$ctx:expr, $($arg:expr),*]) => {
        {
            let ret = $f($ctx.ptr(), $($arg,)*);
            $ctx.check_err().map(|_| ret)
        }
    }
}

pub use check_call;
```
and is called as
```rust
context::check_call!(raw::foo[ctx, other, args, here])
```
I couldn't figure out a way to use `()` for the function call, but I didn't try that hard.

I don't love this, just proposing it as an alternative because I also don't love mini-closures everywhere.

Reply:
We might be over-engineering this.

> * (+) can't accidentally use the same context pointer with multiple raw API functions in a row

:+1: 

> * (+) no overhead of using closures everywhere (not actually sure if there is an overhead of that?)

I assumed it would be inlined.

> * (+) looks like just calling the function

:+1: 

> * (-) macros are janky and magical and give bad error messages

This is important to me.
Also I think editor support won't be as good (rust-analyzer).

Reply:
I've tried it anyway and the editor support was quite decent. The error messages degrade only slightly, as it still reports all relevant positions. They're also slightly harder to understand because of the macro indirection, but it's manageable.
Also these errors aren't "user" facing anyway; library users shouldn't need any of the raw functions, and all wrappers handle this for them.